File: java/01-beginner/04-control-flow/Loops.java
/**
 * LESSON: Loops in Java
 * Repeat blocks of code multiple times.
 */
public class Loops {
    public static void main(String[] args) {
        // ===== FOR LOOP =====
        // for (initialization; condition; update)
        System.out.println("--- For Loop ---");
        for (int i = 1; i <= 5; i++) {
            System.out.println("Count: " + i);
        }

        // ===== WHILE LOOP =====
        // Repeats while condition is true (checks BEFORE each iteration)
        System.out.println("\n--- While Loop ---");
        int n = 5;
        while (n > 0) {
            System.out.println("Countdown: " + n);
            n--;
        }

        // ===== DO-WHILE LOOP =====
        // Executes at least ONCE, then checks condition
        System.out.println("\n--- Do-While Loop ---");
        int num = 1;
        do {
            System.out.println("Number: " + num);
            num++;
        } while (num <= 3);

        // ===== ENHANCED FOR-EACH LOOP =====
        System.out.println("\n--- For-Each Loop ---");
        String[] fruits = {"Apple", "Banana", "Cherry", "Mango"};
        for (String fruit : fruits) {
            System.out.println("Fruit: " + fruit);
        }

        // ===== NESTED LOOPS =====
        System.out.println("\n--- Nested Loops (Multiplication Table) ---");
        for (int i = 1; i <= 3; i++) {
            for (int j = 1; j <= 3; j++) {
                System.out.printf("%d x %d = %d\t", i, j, i * j);
            }
            System.out.println();
        }

        // ===== BREAK =====
        // Exits the loop immediately
        System.out.println("\n--- Break ---");
        for (int i = 1; i <= 10; i++) {
            if (i == 5) {
                System.out.println("Breaking at " + i);
                break;
            }
            System.out.println("i = " + i);
        }

        // ===== CONTINUE =====
        // Skips current iteration, continues with next
        System.out.println("\n--- Continue (skip even numbers) ---");
        for (int i = 1; i <= 10; i++) {
            if (i % 2 == 0) continue; // skip even numbers
            System.out.println("Odd: " + i);
        }

        // ===== LABELED BREAK =====
        // Break out of outer loop from inner loop
        System.out.println("\n--- Labeled Break ---");
        outer:
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (i == 1 && j == 1) {
                    System.out.println("Breaking outer loop at i=" + i + ", j=" + j);
                    break outer;
                }
                System.out.println("i=" + i + ", j=" + j);
            }
        }
    }
}

File: cpp-practice/01-beginner/04-control-flow/loops.cpp
/**
 * LESSON: Loops in C++
 */
#include <iostream>
#include <vector>

int main() {
    // ===== FOR LOOP =====
    std::cout << "--- For Loop ---\n";
    for (int i = 1; i <= 5; i++) {
        std::cout << "Count: " << i << "\n";
    }

    // ===== WHILE LOOP =====
    std::cout << "\n--- While Loop ---\n";
    int n = 5;
    while (n > 0) {
        std::cout << "Countdown: " << n << "\n";
        n--;
    }

    // ===== DO-WHILE LOOP =====
    std::cout << "\n--- Do-While Loop ---\n";
    int num = 1;
    do {
        std::cout << "Number: " << num << "\n";
        num++;
    } while (num <= 3);

    // ===== RANGE-BASED FOR LOOP (C++11) =====
    std::cout << "\n--- Range-Based For ---\n";
    std::vector<std::string> fruits = {"Apple", "Banana", "Cherry", "Mango"};
    for (const auto& fruit : fruits) {
        std::cout << "Fruit: " << fruit << "\n";
    }

    // With arrays
    int numbers[] = {10, 20, 30, 40, 50};
    for (int num : numbers) {
        std::cout << num << " ";
    }
    std::cout << "\n";

    // ===== NESTED LOOPS =====
    std::cout << "\n--- Nested Loops (Multiplication Table) ---\n";
    for (int i = 1; i <= 3; i++) {
        for (int j = 1; j <= 3; j++) {
            std::cout << i << "x" << j << "=" << i * j << "\t";
        }
        std::cout << "\n";
    }

    // ===== BREAK =====
    std::cout << "\n--- Break ---\n";
    for (int i = 1; i <= 10; i++) {
        if (i == 5) {
            std::cout << "Breaking at " << i << "\n";
            break;
        }
        std::cout << "i = " << i << "\n";
    }

    // ===== CONTINUE =====
    std::cout << "\n--- Continue (skip even) ---\n";
    for (int i = 1; i <= 10; i++) {
        if (i % 2 == 0) continue;
        std::cout << "Odd: " << i << "\n";
    }

    return 0;
}
